build_rules.gradle

```diff
@@ -0,0 +1,275 @@
+
+// Provies "htmlDependencyReport" task which spits out a nice page per subproject
+// showing all the dependencies per Gradle configuration.
+apply plugin: "project-report"
+
+// Apply a task dependency visualization plugin which creates a ".dot" file in the build directory
+// giving the task dependencies for that project
+apply plugin: "cz.malohlava.visteg"
+
+/*************************************************************************************************/
+
+/** Used as configuration within the applyJavaNature closure. */
+class JavaNatureConfiguration {
+  double javaVersion = 1.7
+  boolean enableFindbugs = true
+  boolean enableShadow = true
+}
+
+/** Takes a map of parameters. See JavaNatureConfiguration for the set of parameters allowed. */
+ext.applyJavaNature = {
+  println "applyJavaNature with " + (it ? "$it" : "default configuration") + " for project $project.name"
+  // Use the implicit it parameter of the closure to handle zero argument or one argument map calls.
+  JavaNatureConfiguration configuration = it ? it as JavaNatureConfiguration : new JavaNatureConfiguration()
+  apply plugin: "maven"
+  apply plugin: "java"
+
+  sourceCompatibility = configuration.javaVersion
+  targetCompatibility = configuration.javaVersion
+  tasks.withType(JavaCompile) {
+    options.encoding = "UTF-8"
+  }
+
+  test {
+    // Configure the gradle include defaults to be equivalent to the maven-surefire-plugin
+    // http://maven.apache.org/components/surefire/maven-surefire-plugin/test-mojo.html
+    include "**/Test*.class"
+    include "**/*Test.class"
+    include "**/*Tests.class"
+    include "**/*TestCase.class"
+    useJUnit { }
+  }
+
+  apply plugin: "net.ltgt.apt"
+  dependencies {
+    def auto_value = "com.google.auto.value:auto-value:1.5.1"
+    def auto_service = "com.google.auto.service:auto-service:1.0-rc2"
+    compileOnly auto_value
+    apt auto_value
+    testCompileOnly auto_value
+    testApt auto_value
+
+    compileOnly auto_service
+    apt auto_service
+    testCompileOnly auto_service
+    testApt auto_service
+  }
+
+  // Add the optional and provided scopes for dependencies
+  apply plugin: 'propdeps'
+  apply plugin: 'propdeps-maven'
+
+  apply plugin: 'checkstyle'
+
+  tasks.withType(Checkstyle) {
+    configFile = project(":").file("sdks/java/build-tools/src/main/resources/beam/checkstyle.xml")
+    configProperties = [ "checkstyle.suppressions.file": project(":").file("sdks/java/build-tools/src/main/resources/beam/suppressions.xml") ]
+    showViolations = true
+    maxErrors = 0
+  }
+
+  apply plugin: 'com.diffplug.gradle.spotless'
+  spotless {
+    java {
+      target rootProject.fileTree(rootProject.rootDir) {
+        include 'sdks/java/**/*.java'
+      }
+      // Code formatting disabled because style rules are out of date.
+      // eclipse().configFile(rootProject.file('sdks/java/build-tools/src/main/resources/beam/beam-codestyle.xml'))
+    }
+  }
+
+  if (configuration.enableFindbugs) {
```

Comment:
<!--new_thread; commit:8ca3fdca276a74973cd428d7321b995f7b7ebccf; resolved:0-->
Why `if`? Can it be always on and be a separate target? It would be preferable to manage all aspects of what runs by specifying targets. I know in Gradle these are "tasks" but aren't those about the same?

Reply:
Applying a plugin typically adds configuration sets, enables tasks to execute. I have an `if` here to prevent the plugin from applying for a few modules (google-cloud-dataflow-java, google-cloud-platform, kafka, model/*) mirroring the Maven configuration.

Gradle doesn't follow the Maven plugin style where there is a `skip` flag so either you enable the plugin or you don't. Some plugins have the ability to not fail the build if they fail.